onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlDFT.h

```diff
@@ -0,0 +1,695 @@
+        ID3D12Resource* /*outputResource*/,
+        ID3D12GraphicsCommandList* /*commandList*/)
+    {
+        ORT_THROW_HR(E_NOTIMPL);
+    }
+};
+
+struct DFTShapeInferrer : public WRL::Base<IMLOperatorShapeInferrer>
+{
+    STDMETHOD(InferOutputShapes)(IMLOperatorShapeInferenceContext* context) noexcept
+    {
+        try
+        {
+            int64_t axis;
+            ORT_THROW_IF_FAILED(context->GetAttribute("axis", MLOperatorAttributeType::Int, 1, sizeof(int64_t), reinterpret_cast<void*>(&axis)));
+            int64_t isInverseInt;
+            ORT_THROW_IF_FAILED(context->GetAttribute("inverse", MLOperatorAttributeType::Int, 1, sizeof(int64_t), reinterpret_cast<void*>(&isInverseInt)));
+            int64_t isOnesidedInt;
+            ORT_THROW_IF_FAILED(context->GetAttribute("onesided", MLOperatorAttributeType::Int, 1, sizeof(int64_t), reinterpret_cast<void*>(&isOnesidedInt)));
+            bool isOnesided = static_cast<bool>(isOnesidedInt);
+            bool inverse = static_cast<bool>(isInverseInt);
+
+            if (inverse && isOnesided) {
+                throw new std::exception("isOnesided and inverse attributes cannot be enabled at the same time");
+            }
+
+            uint32_t rank;
+            ORT_THROW_IF_FAILED(context->GetInputTensorDimensionCount(0, &rank));
+            if (rank == 0) {
+                // If no shape is available for the input, skip shape inference...
+                throw;
+            }
+
+            auto axisIdx = OperatorHelper::HandleNegativeAxis(static_cast<int32_t>(axis), rank);
+
+            // if (context->IsInputValid(1))
+            // {
+            //     // If dft_length is specified, then we should honor the shape.
+            //     // Set the output dimension to match the dft_length on the axis.
+            //     // If onesided this will be adjusted later on.
+            //     // However, this is not supported as of now, so do nothing...
+            // }
+
+            // In general the output shape will match the input shape exactly
+            // So initialize the output shape with the input shape
+            std::vector<uint32_t> inputDims(rank);
+            ORT_THROW_IF_FAILED(context->GetInputTensorShape(0, rank, inputDims.data()));
+            auto outputDims = inputDims;
+            outputDims.back() = 2;
+
+            // When DFT is onesided, the output shape is half the size of the input shape
+            // along the specified axis.
+            if (isOnesided) {
+                auto axisDimension = outputDims.at(axisIdx);
+                // We need to update the output shape dimension along the specified axis,
+                // but sometimes the dimension will be a free dimension or be otherwise unset.
+                // Only perform inference when a input dimension value exists.
+                auto originalSignalSize = axisDimension;
+                auto halfSignalSize = (originalSignalSize >> 1) + 1;
```

Comment:
>(originalSignalSize >> 1) + 1 [](http://example.com/codeflow?start=38&length=29)

Should this be ((originalSignalSize + 1) >> 1), if your intention is to halve with round up?

Otherwise `originalSignalSize = 32` will yield `halfSignalSize = 17`, not 16 🤨.

Reply:
Yep it should be 17 :)